core/appcontainer/MessageModel.js

```diff
@@ -61,15 +51,20 @@ export class MessageModel {
     @action
     doConfirm() {
         if (this.onConfirm) this.onConfirm();
+        this._resolver(true);
         this.close();
     }
 
     @action
     doCancel() {
         if (this.onCancel) this.onCancel();
+        this._resolver(false);
```

Comment:
For my specific use-case where I want to return an `XH.confirm()`  in a router5 `canDeactivate` callback, I would need the Promise to be rejected in order to prevent navigation. Just tried it in Apollo and resolving the promise with `false` will not prevent the navigation. I think I've seen both styles of using Promises with different dialog libraries, rejecting on cancel or passing the result in resolve, not sure which one I think makes the most sense from a API design perspective. If we do keep this approach, it will require a bit more code in my use-case to call the `done()` callback that router5 passes to the `canDeactivate` callback based on the result, so wouldn't actually clean up my code where I am currently using `onConfirm`. Thoughts?

Reply:
Can we look at your code together again?  Definitely want to understand it -- it seems like it may be more complicated than the typical use case.

Logically, it seems more appropriate to resolve to a boolean than to throw.  Its really not an "exception" or an "error" -- it seems to me that cancel is  a totally expected and common code path that ought to be symmetrical with confirm.

Reply:
Yeah it is definitely different from the typical use-case, and very specific to the router5 API. 100% agree with your point on resolving to a boolean. Happy to look at it with you again, I think #459 is the more appropriate place to cleanup my use-case now that I've thought about it more (perhaps returning a string from the `canDeactivate` hook would result in a confirm prompt to minimize the amount of wiring apps would need to do)